core/src/main/java/org/mule/enricher/MessageEnricher.java

```diff
@@ -63,9 +67,13 @@ public MuleEvent process(MuleEvent event) throws MuleException
     {
         ExpressionManager expressionManager = event.getMuleContext().getExpressionManager();
 
-        MuleEvent enricherEvent;
         //TODO: change DefaultMuleEvent.copy to DefaultMuleEvent.copyPreservingSession
-        enricherEvent = DefaultMuleEvent.copy(event);
+        MuleEvent enricherEvent = DefaultMuleEvent.copy(event);
+        if (event.isAllowNonBlocking() && event.getReplyToHandler() != null)
+        {
+            // TODO: MULE-8662 Support non-blocking execution within message enricher
+            enricherEvent = new DefaultMuleEvent(event.getMessage(), event, true);
```

Comment:
Then add a TODO comment to fix this when we have a builder for MuleEvent. I think the could be avoided now by adding an overload to DefaultMuleEvent.copy but is an ugly solution...